File: electron/models/EleveData.js
import { v4 as uuidv4 } from 'uuid';

export class EleveData {
  constructor(data = {}) {
    this.id = data.id || uuidv4();
    this.nom = data.nom || '';
    this.prenom = data.prenom || '';
    this.dateNaissance = data.dateNaissance || data.date_naissance || '';
    this.sexe = data.sexe || 'M';
    this.photo = data.photo || null;
    this.anneeScolaire = data.anneeScolaire || data.annee_scolaire || '';
    this.matricule = data.matricule || '';
    this.statut = data.statut || 'actif';
    this.lieuDeNaissance = data.lieuDeNaissance || data.lieu_de_naissance || null;
    this.contact = data.contact || null;
    this.createdAt = data.createdAt || data.created_at || new Date().toISOString();
    this.updatedAt = data.updatedAt || data.updated_at || new Date().toISOString();
  }

  toDB() {
    return {
      id: this.id,
      nom: this.nom,
      prenom: this.prenom,
      date_naissance: this.dateNaissance,
      sexe: this.sexe,
      photo: this.photo,
      annee_scolaire: this.anneeScolaire,
      matricule: this.matricule,
      statut: this.statut,
      lieu_de_naissance: this.lieuDeNaissance,
      contact: this.contact,
      created_at: this.createdAt,
      updated_at: this.updatedAt
    };
  }

  toJSON() {
    return {
      id: this.id,
      nom: this.nom,
      prenom: this.prenom,
      dateNaissance: this.dateNaissance,
      sexe: this.sexe,
      photo: this.photo,
      anneeScolaire: this.anneeScolaire,
      matricule: this.matricule,
      statut: this.statut,
      lieuDeNaissance: this.lieuDeNaissance,
      contact: this.contact,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }

  validate() {
    const errors = [];
    if (!this.nom) errors.push('Le nom est requis');
    if (!this.prenom) errors.push('Le prénom est requis');
    if (!this.dateNaissance) errors.push('La date de naissance est requise');
    if (!this.matricule) errors.push('Le matricule est requis');
    if (!['M', 'F'].includes(this.sexe)) errors.push('Le sexe doit être M ou F');
    if (!['actif', 'inactif', 'exclu'].includes(this.statut)) {
      errors.push('Le statut doit être actif, inactif ou exclu');
    }
    return errors;
  }
}

File: src-tauri/src/models/eleve_data.rs
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};
use uuid::Uuid;
use sqlx::Type;
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, Type, FromRow)]
#[sqlx(type_name = "TEXT", rename_all = "lowercase")]
pub enum Sexe {
    #[serde(rename = "M")]
    M,
    #[serde(rename = "F")]
    F,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, FromRow)]
#[sqlx(type_name = "TEXT", rename_all = "lowercase")]
pub enum StatutEleve {
    #[serde(rename = "actif")]
    Actif,
    #[serde(rename = "inactif")]
    Inactif,
    #[serde(rename = "exclu")]
    Exclu,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BaseEleveData {
    pub nom: String,
    pub prenom: String,
    pub date_naissance: String,
    pub sexe: Sexe,
    pub photo: Option<String>,
    pub annee_scolaire: String,
    pub matricule: String,
    pub statut: StatutEleve,
    pub lieu_de_naissance: Option<String>,
    pub contact: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct EleveData {
    pub id: String,
    pub nom: String,
    pub prenom: String,
    pub date_naissance: String,
    pub sexe: Sexe,
    pub photo: Option<String>,
    pub annee_scolaire: String,
    pub matricule: String,
    pub statut: StatutEleve,
    pub lieu_de_naissance: Option<String>,
    pub contact: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

impl From<BaseEleveData> for EleveData {
    fn from(base: BaseEleveData) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::new_v4().to_string(),
            nom: base.nom,
            prenom: base.prenom,
            date_naissance: base.date_naissance,
            sexe: base.sexe,
            photo: base.photo,
            annee_scolaire: base.annee_scolaire,
            matricule: base.matricule,
            statut: base.statut,
            lieu_de_naissance: base.lieu_de_naissance,
            contact: base.contact,
            created_at: Some(now),
            updated_at: Some(now),
        }
    }
}